# Build a Regression Model in Keras


The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:
1. Cement
2. Blast Furnace Slag
3. Fly Ash
4. Water
5. Superplasticizer
6. Coarse Aggregate
7. Fine Aggregate

The target variable in this problem is the concrete sample strength.

In [1]:
import pandas as pd
import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense

from sklearn.model_selection import train_test_split



## ImportDataset


In [2]:
# First let's import the data in to a pandas dataframe
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
# Let's check the dataset for any missing values.

concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [4]:
# Split data into predictors and target

concrete_data_columns = concrete_data.columns

X = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
y = concrete_data['Strength'] # Strength column

# Normalize the data by substracting the mean and dividing by the standard deviation.
X_norm = (X - X.mean()) / X.std()
X_norm.head()

n_cols = X_norm.shape[1] # number of predictors

## Build the Neural Network

In [5]:
# define regression model

def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

### Train and Test the Network

In [6]:

# create a list of 50 mean squared errors.

list_mse = np.zeros(50)

for i in range(0,50):

    #Split the data in train and test dataset
    X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=0.3, random_state=4)

    # build the model
    model = regression_model()

    # fit the model
    model.fit(X_norm, y, epochs=50, verbose=2)

    # evaluate the model
    scores = model.evaluate(X_test, y_test, verbose=0)
    
    list_mse[i] = scores


Epoch 1/50
33/33 - 0s - loss: 1589.9341
Epoch 2/50
33/33 - 0s - loss: 1550.0490
Epoch 3/50
33/33 - 0s - loss: 1504.1809
Epoch 4/50
33/33 - 0s - loss: 1431.9736
Epoch 5/50
33/33 - 0s - loss: 1310.3137
Epoch 6/50
33/33 - 0s - loss: 1116.6051
Epoch 7/50
33/33 - 0s - loss: 849.4273
Epoch 8/50
33/33 - 0s - loss: 562.6445
Epoch 9/50
33/33 - 0s - loss: 367.6833
Epoch 10/50
33/33 - 0s - loss: 281.5534
Epoch 11/50
33/33 - 0s - loss: 244.3721
Epoch 12/50
33/33 - 0s - loss: 220.7040
Epoch 13/50
33/33 - 0s - loss: 204.2919
Epoch 14/50
33/33 - 0s - loss: 192.0873
Epoch 15/50
33/33 - 0s - loss: 183.0368
Epoch 16/50
33/33 - 0s - loss: 175.0979
Epoch 17/50
33/33 - 0s - loss: 169.3129
Epoch 18/50
33/33 - 0s - loss: 164.4589
Epoch 19/50
33/33 - 0s - loss: 160.9549
Epoch 20/50
33/33 - 0s - loss: 157.0411
Epoch 21/50
33/33 - 0s - loss: 153.9728
Epoch 22/50
33/33 - 0s - loss: 151.2873
Epoch 23/50
33/33 - 0s - loss: 148.4461
Epoch 24/50
33/33 - 0s - loss: 146.2728
Epoch 25/50
33/33 - 0s - loss: 144.0208
Epo

Epoch 8/50
33/33 - 0s - loss: 514.4536
Epoch 9/50
33/33 - 0s - loss: 344.1288
Epoch 10/50
33/33 - 0s - loss: 276.3894
Epoch 11/50
33/33 - 0s - loss: 243.6879
Epoch 12/50
33/33 - 0s - loss: 222.7629
Epoch 13/50
33/33 - 0s - loss: 206.8629
Epoch 14/50
33/33 - 0s - loss: 194.3601
Epoch 15/50
33/33 - 0s - loss: 184.3796
Epoch 16/50
33/33 - 0s - loss: 176.4783
Epoch 17/50
33/33 - 0s - loss: 168.8515
Epoch 18/50
33/33 - 0s - loss: 163.2469
Epoch 19/50
33/33 - 0s - loss: 158.6346
Epoch 20/50
33/33 - 0s - loss: 153.8065
Epoch 21/50
33/33 - 0s - loss: 149.7207
Epoch 22/50
33/33 - 0s - loss: 145.5341
Epoch 23/50
33/33 - 0s - loss: 141.7949
Epoch 24/50
33/33 - 0s - loss: 138.3362
Epoch 25/50
33/33 - 0s - loss: 134.7211
Epoch 26/50
33/33 - 0s - loss: 131.6590
Epoch 27/50
33/33 - 0s - loss: 127.1031
Epoch 28/50
33/33 - 0s - loss: 123.8109
Epoch 29/50
33/33 - 0s - loss: 120.1087
Epoch 30/50
33/33 - 0s - loss: 117.1284
Epoch 31/50
33/33 - 0s - loss: 113.6993
Epoch 32/50
33/33 - 0s - loss: 110.4361
Ep

Epoch 14/50
33/33 - 0s - loss: 196.2444
Epoch 15/50
33/33 - 0s - loss: 190.1265
Epoch 16/50
33/33 - 0s - loss: 185.2972
Epoch 17/50
33/33 - 0s - loss: 179.9315
Epoch 18/50
33/33 - 0s - loss: 175.7457
Epoch 19/50
33/33 - 0s - loss: 171.4924
Epoch 20/50
33/33 - 0s - loss: 167.5523
Epoch 21/50
33/33 - 0s - loss: 163.8391
Epoch 22/50
33/33 - 0s - loss: 160.5050
Epoch 23/50
33/33 - 0s - loss: 157.0695
Epoch 24/50
33/33 - 0s - loss: 154.1190
Epoch 25/50
33/33 - 0s - loss: 150.4884
Epoch 26/50
33/33 - 0s - loss: 147.6268
Epoch 27/50
33/33 - 0s - loss: 144.7110
Epoch 28/50
33/33 - 0s - loss: 142.2310
Epoch 29/50
33/33 - 0s - loss: 139.8723
Epoch 30/50
33/33 - 0s - loss: 138.4777
Epoch 31/50
33/33 - 0s - loss: 136.0616
Epoch 32/50
33/33 - 0s - loss: 134.5989
Epoch 33/50
33/33 - 0s - loss: 133.0734
Epoch 34/50
33/33 - 0s - loss: 131.8881
Epoch 35/50
33/33 - 0s - loss: 130.4162
Epoch 36/50
33/33 - 0s - loss: 129.1916
Epoch 37/50
33/33 - 0s - loss: 128.0145
Epoch 38/50
33/33 - 0s - loss: 127.0472


Epoch 20/50
33/33 - 0s - loss: 149.8621
Epoch 21/50
33/33 - 0s - loss: 148.6795
Epoch 22/50
33/33 - 0s - loss: 146.1301
Epoch 23/50
33/33 - 0s - loss: 144.4870
Epoch 24/50
33/33 - 0s - loss: 143.2881
Epoch 25/50
33/33 - 0s - loss: 141.9500
Epoch 26/50
33/33 - 0s - loss: 140.6984
Epoch 27/50
33/33 - 0s - loss: 139.4788
Epoch 28/50
33/33 - 0s - loss: 138.1725
Epoch 29/50
33/33 - 0s - loss: 136.7451
Epoch 30/50
33/33 - 0s - loss: 135.5045
Epoch 31/50
33/33 - 0s - loss: 134.8246
Epoch 32/50
33/33 - 0s - loss: 133.5715
Epoch 33/50
33/33 - 0s - loss: 132.8337
Epoch 34/50
33/33 - 0s - loss: 132.1172
Epoch 35/50
33/33 - 0s - loss: 130.2180
Epoch 36/50
33/33 - 0s - loss: 129.2386
Epoch 37/50
33/33 - 0s - loss: 128.4045
Epoch 38/50
33/33 - 0s - loss: 127.0085
Epoch 39/50
33/33 - 0s - loss: 125.9929
Epoch 40/50
33/33 - 0s - loss: 125.2483
Epoch 41/50
33/33 - 0s - loss: 124.2157
Epoch 42/50
33/33 - 0s - loss: 123.2107
Epoch 43/50
33/33 - 0s - loss: 122.3992
Epoch 44/50
33/33 - 0s - loss: 121.0420


Epoch 26/50
33/33 - 0s - loss: 144.9816
Epoch 27/50
33/33 - 0s - loss: 141.0552
Epoch 28/50
33/33 - 0s - loss: 137.6355
Epoch 29/50
33/33 - 0s - loss: 134.1259
Epoch 30/50
33/33 - 0s - loss: 130.4980
Epoch 31/50
33/33 - 0s - loss: 126.8206
Epoch 32/50
33/33 - 0s - loss: 122.8257
Epoch 33/50
33/33 - 0s - loss: 118.6373
Epoch 34/50
33/33 - 0s - loss: 114.8223
Epoch 35/50
33/33 - 0s - loss: 111.1138
Epoch 36/50
33/33 - 0s - loss: 107.7011
Epoch 37/50
33/33 - 0s - loss: 104.3641
Epoch 38/50
33/33 - 0s - loss: 100.2489
Epoch 39/50
33/33 - 0s - loss: 96.9312
Epoch 40/50
33/33 - 0s - loss: 93.5880
Epoch 41/50
33/33 - 0s - loss: 90.1973
Epoch 42/50
33/33 - 0s - loss: 87.0956
Epoch 43/50
33/33 - 0s - loss: 84.4987
Epoch 44/50
33/33 - 0s - loss: 81.5309
Epoch 45/50
33/33 - 0s - loss: 79.0303
Epoch 46/50
33/33 - 0s - loss: 76.7596
Epoch 47/50
33/33 - 0s - loss: 73.9810
Epoch 48/50
33/33 - 0s - loss: 72.3126
Epoch 49/50
33/33 - 0s - loss: 70.2758
Epoch 50/50
33/33 - 0s - loss: 68.9167
Epoch 1/50
3

Epoch 28/50
33/33 - 0s - loss: 136.1497
Epoch 29/50
33/33 - 0s - loss: 134.5914
Epoch 30/50
33/33 - 0s - loss: 133.6946
Epoch 31/50
33/33 - 0s - loss: 132.3713
Epoch 32/50
33/33 - 0s - loss: 131.7151
Epoch 33/50
33/33 - 0s - loss: 130.6373
Epoch 34/50
33/33 - 0s - loss: 129.4290
Epoch 35/50
33/33 - 0s - loss: 128.8065
Epoch 36/50
33/33 - 0s - loss: 127.4325
Epoch 37/50
33/33 - 0s - loss: 126.6014
Epoch 38/50
33/33 - 0s - loss: 125.6046
Epoch 39/50
33/33 - 0s - loss: 125.1401
Epoch 40/50
33/33 - 0s - loss: 124.1581
Epoch 41/50
33/33 - 0s - loss: 123.4053
Epoch 42/50
33/33 - 0s - loss: 122.5036
Epoch 43/50
33/33 - 0s - loss: 121.7207
Epoch 44/50
33/33 - 0s - loss: 121.1830
Epoch 45/50
33/33 - 0s - loss: 120.2589
Epoch 46/50
33/33 - 0s - loss: 119.2812
Epoch 47/50
33/33 - 0s - loss: 118.7653
Epoch 48/50
33/33 - 0s - loss: 118.2161
Epoch 49/50
33/33 - 0s - loss: 117.4033
Epoch 50/50
33/33 - 0s - loss: 116.7501
Epoch 1/50
33/33 - 0s - loss: 1556.6095
Epoch 2/50
33/33 - 0s - loss: 1528.9802


Epoch 34/50
33/33 - 0s - loss: 96.3616
Epoch 35/50
33/33 - 0s - loss: 94.5607
Epoch 36/50
33/33 - 0s - loss: 93.4974
Epoch 37/50
33/33 - 0s - loss: 92.2200
Epoch 38/50
33/33 - 0s - loss: 91.0156
Epoch 39/50
33/33 - 0s - loss: 89.7478
Epoch 40/50
33/33 - 0s - loss: 88.6318
Epoch 41/50
33/33 - 0s - loss: 87.2441
Epoch 42/50
33/33 - 0s - loss: 86.2967
Epoch 43/50
33/33 - 0s - loss: 85.2114
Epoch 44/50
33/33 - 0s - loss: 84.0720
Epoch 45/50
33/33 - 0s - loss: 82.8794
Epoch 46/50
33/33 - 0s - loss: 81.9330
Epoch 47/50
33/33 - 0s - loss: 81.0548
Epoch 48/50
33/33 - 0s - loss: 80.3513
Epoch 49/50
33/33 - 0s - loss: 79.2167
Epoch 50/50
33/33 - 0s - loss: 78.0751
Epoch 1/50
33/33 - 0s - loss: 1529.9705
Epoch 2/50
33/33 - 0s - loss: 1474.6770
Epoch 3/50
33/33 - 0s - loss: 1388.4948
Epoch 4/50
33/33 - 0s - loss: 1236.0563
Epoch 5/50
33/33 - 0s - loss: 987.8749
Epoch 6/50
33/33 - 0s - loss: 687.2581
Epoch 7/50
33/33 - 0s - loss: 438.2321
Epoch 8/50
33/33 - 0s - loss: 288.3336
Epoch 9/50
33/33 - 0s

Epoch 40/50
33/33 - 0s - loss: 97.7720
Epoch 41/50
33/33 - 0s - loss: 96.0034
Epoch 42/50
33/33 - 0s - loss: 94.1241
Epoch 43/50
33/33 - 0s - loss: 93.5135
Epoch 44/50
33/33 - 0s - loss: 91.1303
Epoch 45/50
33/33 - 0s - loss: 90.5817
Epoch 46/50
33/33 - 0s - loss: 88.6829
Epoch 47/50
33/33 - 0s - loss: 87.4348
Epoch 48/50
33/33 - 0s - loss: 86.1237
Epoch 49/50
33/33 - 0s - loss: 85.0352
Epoch 50/50
33/33 - 0s - loss: 84.0093
Epoch 1/50
33/33 - 0s - loss: 1507.6797
Epoch 2/50
33/33 - 0s - loss: 1444.2649
Epoch 3/50
33/33 - 0s - loss: 1327.8789
Epoch 4/50
33/33 - 0s - loss: 1122.3636
Epoch 5/50
33/33 - 0s - loss: 833.3038
Epoch 6/50
33/33 - 0s - loss: 522.6458
Epoch 7/50
33/33 - 0s - loss: 309.9548
Epoch 8/50
33/33 - 0s - loss: 226.2393
Epoch 9/50
33/33 - 0s - loss: 204.5261
Epoch 10/50
33/33 - 0s - loss: 194.1370
Epoch 11/50
33/33 - 0s - loss: 187.0965
Epoch 12/50
33/33 - 0s - loss: 180.2599
Epoch 13/50
33/33 - 0s - loss: 174.7465
Epoch 14/50
33/33 - 0s - loss: 168.9726
Epoch 15/50
33/3

Epoch 47/50
33/33 - 0s - loss: 118.9827
Epoch 48/50
33/33 - 0s - loss: 118.5580
Epoch 49/50
33/33 - 0s - loss: 117.5814
Epoch 50/50
33/33 - 0s - loss: 117.0311
Epoch 1/50
33/33 - 0s - loss: 1535.6187
Epoch 2/50
33/33 - 0s - loss: 1475.5608
Epoch 3/50
33/33 - 0s - loss: 1373.3979
Epoch 4/50
33/33 - 0s - loss: 1205.7289
Epoch 5/50
33/33 - 0s - loss: 963.3034
Epoch 6/50
33/33 - 0s - loss: 686.4271
Epoch 7/50
33/33 - 0s - loss: 456.7038
Epoch 8/50
33/33 - 0s - loss: 306.3394
Epoch 9/50
33/33 - 0s - loss: 227.0451
Epoch 10/50
33/33 - 0s - loss: 191.2502
Epoch 11/50
33/33 - 0s - loss: 176.0177
Epoch 12/50
33/33 - 0s - loss: 167.9497
Epoch 13/50
33/33 - 0s - loss: 162.9519
Epoch 14/50
33/33 - 0s - loss: 159.0092
Epoch 15/50
33/33 - 0s - loss: 154.9406
Epoch 16/50
33/33 - 0s - loss: 152.3694
Epoch 17/50
33/33 - 0s - loss: 149.6501
Epoch 18/50
33/33 - 0s - loss: 147.1912
Epoch 19/50
33/33 - 0s - loss: 144.8459
Epoch 20/50
33/33 - 0s - loss: 143.0651
Epoch 21/50
33/33 - 0s - loss: 140.7020
Epoch

Epoch 4/50
33/33 - 0s - loss: 1184.5717
Epoch 5/50
33/33 - 0s - loss: 884.2572
Epoch 6/50
33/33 - 0s - loss: 555.9334
Epoch 7/50
33/33 - 0s - loss: 344.8828
Epoch 8/50
33/33 - 0s - loss: 272.4752
Epoch 9/50
33/33 - 0s - loss: 244.5659
Epoch 10/50
33/33 - 0s - loss: 226.6269
Epoch 11/50
33/33 - 0s - loss: 212.3833
Epoch 12/50
33/33 - 0s - loss: 200.3853
Epoch 13/50
33/33 - 0s - loss: 189.9749
Epoch 14/50
33/33 - 0s - loss: 180.3546
Epoch 15/50
33/33 - 0s - loss: 171.6913
Epoch 16/50
33/33 - 0s - loss: 164.3267
Epoch 17/50
33/33 - 0s - loss: 157.1158
Epoch 18/50
33/33 - 0s - loss: 150.7308
Epoch 19/50
33/33 - 0s - loss: 144.7955
Epoch 20/50
33/33 - 0s - loss: 139.0612
Epoch 21/50
33/33 - 0s - loss: 134.2202
Epoch 22/50
33/33 - 0s - loss: 129.1997
Epoch 23/50
33/33 - 0s - loss: 124.3817
Epoch 24/50
33/33 - 0s - loss: 120.5911
Epoch 25/50
33/33 - 0s - loss: 116.9545
Epoch 26/50
33/33 - 0s - loss: 114.0761
Epoch 27/50
33/33 - 0s - loss: 110.9047
Epoch 28/50
33/33 - 0s - loss: 108.4306
Epoch

Epoch 11/50
33/33 - 0s - loss: 190.9030
Epoch 12/50
33/33 - 0s - loss: 176.8774
Epoch 13/50
33/33 - 0s - loss: 169.0980
Epoch 14/50
33/33 - 0s - loss: 164.4955
Epoch 15/50
33/33 - 0s - loss: 161.1812
Epoch 16/50
33/33 - 0s - loss: 158.2139
Epoch 17/50
33/33 - 0s - loss: 155.3889
Epoch 18/50
33/33 - 0s - loss: 153.0763
Epoch 19/50
33/33 - 0s - loss: 150.2437
Epoch 20/50
33/33 - 0s - loss: 146.9824
Epoch 21/50
33/33 - 0s - loss: 144.2944
Epoch 22/50
33/33 - 0s - loss: 140.4456
Epoch 23/50
33/33 - 0s - loss: 137.4753
Epoch 24/50
33/33 - 0s - loss: 134.0611
Epoch 25/50
33/33 - 0s - loss: 130.1644
Epoch 26/50
33/33 - 0s - loss: 126.8245
Epoch 27/50
33/33 - 0s - loss: 123.9218
Epoch 28/50
33/33 - 0s - loss: 120.4736
Epoch 29/50
33/33 - 0s - loss: 117.8385
Epoch 30/50
33/33 - 0s - loss: 114.9340
Epoch 31/50
33/33 - 0s - loss: 112.1033
Epoch 32/50
33/33 - 0s - loss: 110.0455
Epoch 33/50
33/33 - 0s - loss: 107.9346
Epoch 34/50
33/33 - 0s - loss: 105.7738
Epoch 35/50
33/33 - 0s - loss: 103.9274


Epoch 13/50
33/33 - 0s - loss: 181.1807
Epoch 14/50
33/33 - 0s - loss: 176.5398
Epoch 15/50
33/33 - 0s - loss: 171.9570
Epoch 16/50
33/33 - 0s - loss: 168.2249
Epoch 17/50
33/33 - 0s - loss: 165.0953
Epoch 18/50
33/33 - 0s - loss: 162.1234
Epoch 19/50
33/33 - 0s - loss: 158.9135
Epoch 20/50
33/33 - 0s - loss: 155.7977
Epoch 21/50
33/33 - 0s - loss: 153.2644
Epoch 22/50
33/33 - 0s - loss: 151.1853
Epoch 23/50
33/33 - 0s - loss: 148.9215
Epoch 24/50
33/33 - 0s - loss: 146.6445
Epoch 25/50
33/33 - 0s - loss: 144.7936
Epoch 26/50
33/33 - 0s - loss: 142.5865
Epoch 27/50
33/33 - 0s - loss: 140.6918
Epoch 28/50
33/33 - 0s - loss: 139.0921
Epoch 29/50
33/33 - 0s - loss: 137.4398
Epoch 30/50
33/33 - 0s - loss: 135.7652
Epoch 31/50
33/33 - 0s - loss: 134.5224
Epoch 32/50
33/33 - 0s - loss: 132.8403
Epoch 33/50
33/33 - 0s - loss: 131.5741
Epoch 34/50
33/33 - 0s - loss: 130.1702
Epoch 35/50
33/33 - 0s - loss: 128.6803
Epoch 36/50
33/33 - 0s - loss: 127.0387
Epoch 37/50
33/33 - 0s - loss: 125.8193


Epoch 19/50
33/33 - 0s - loss: 168.9931
Epoch 20/50
33/33 - 0s - loss: 164.8149
Epoch 21/50
33/33 - 0s - loss: 160.7358
Epoch 22/50
33/33 - 0s - loss: 157.4137
Epoch 23/50
33/33 - 0s - loss: 153.9449
Epoch 24/50
33/33 - 0s - loss: 150.7562
Epoch 25/50
33/33 - 0s - loss: 147.5391
Epoch 26/50
33/33 - 0s - loss: 145.1650
Epoch 27/50
33/33 - 0s - loss: 142.5161
Epoch 28/50
33/33 - 0s - loss: 140.7781
Epoch 29/50
33/33 - 0s - loss: 138.4991
Epoch 30/50
33/33 - 0s - loss: 136.6709
Epoch 31/50
33/33 - 0s - loss: 134.7098
Epoch 32/50
33/33 - 0s - loss: 132.8002
Epoch 33/50
33/33 - 0s - loss: 131.4080
Epoch 34/50
33/33 - 0s - loss: 129.4217
Epoch 35/50
33/33 - 0s - loss: 127.7454
Epoch 36/50
33/33 - 0s - loss: 126.0630
Epoch 37/50
33/33 - 0s - loss: 124.6695
Epoch 38/50
33/33 - 0s - loss: 123.0336
Epoch 39/50
33/33 - 0s - loss: 121.7539
Epoch 40/50
33/33 - 0s - loss: 119.9195
Epoch 41/50
33/33 - 0s - loss: 119.2494
Epoch 42/50
33/33 - 0s - loss: 117.3207
Epoch 43/50
33/33 - 0s - loss: 116.3467


## Report the mean and the standard deviation of the mean squared errors.

In [7]:
print("Mean of the 50 MSE list: ", np.mean(list_mse))

print("Standard Deviation of the 50 MSE list: ", np.std(list_mse))


Mean of the 50 MSE list:  107.76124275207519
Standard Deviation of the 50 MSE list:  26.140556515617316


How does the mean of the mean squared errors compare to that from Step B?

|  Part |    MSE   |   STD   |
| ----- |  ------- | ------- |
|   A   |  245.98  |  277.59 |
|   B   |  220.83  |  34.17  |
|   C   |  134.85  |  10.36  |
|   D   |  107.76  |  26.14  |
